lib/logstash/filters/truncate.rb

```diff
@@ -0,0 +1,92 @@
+# encoding: utf-8
+require "logstash/filters/base"
+require "logstash/namespace"
+
+# Allows you to truncate fields longer than a given length.
+#
+# This truncates on bytes values, not character count.  In practice, this
+# should mean that the truncated length is somewhere between `length_bytes` and
+# `length_bytes - 6` (UTF-8 supports up to 6-byte characters).
+class LogStash::Filters::Truncate < LogStash::Filters::Base
+  config_name "truncate"
+
+  # A list of fieldrefs to truncate if they are too long.
+  #
+  # If not specified, the default behavior will be to attempt truncation on all
+  # strings in the event. This default behavior could be computationally
+  # expensive, so if you know exactly which fields you wish to truncate, it is
+  # advised that you be specific and configure the fields you want truncated.
+  config :fields, :validate => :string, :list => true
+
+  # Fields over this length will be truncated to this length.
+  #
+  # Truncation happens from the end of the text (the start will be kept).
+  #
+  # As an example, if you set `length_bytes => 10` and a field contains "hello
+  # world, how are you?", then this field will be truncated and have this value:
+  # "hello worl"
+  config :length_bytes, :validate => :number, :required => true
+
+  def register
+    # nothing
+  end
+
+  def filter(event)
+    if @fields
+      @fields.each do |field|
+        truncate(event, field, @length_bytes)
+      end
+    else
+      truncate_all(event, @length_bytes)
+    end
+  end
+
+  private
+  def truncate(event, field, length)
+    value = event.get(field)
+    if value.is_a?(String)
+      event.set(field, trim(value, length))
+    elsif value.is_a?(Array)
+      new_list = value.map { |v| v.is_a?(String) ? trim(v, length) : v }
+      event.set(field, new_list)
+    elsif value.is_a?(Hash) || value.is_a?(java.util.Map)
+      value.keys.each do |key|
+        truncate(event, "#{field}[#{key}]", length)
+      end
+    end
+  end
```

Comment:
These probably belong as utility methods somewhere else. I'll make them be class methods instead.